### PR TITLE
At most one flow/factor states calculation for one contingency in DC sensitivity analysis

### PR DESCRIPTION
**Pull request by p-arvy**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No.

**What kind of change does this PR introduce?**

**What is the current behavior?**

Multiple modified flow states calculations can be performed for a given contigency in DC sensitivity analysis, e.g. when the connectivity changed and a pst is lost. Same for factor states.

**What is the new behavior (if this is a feature change)?**
At most one flow/factor states calculation is performed for each contingency.

**Does this PR introduce a breaking change or deprecate an API?**
- [ ] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**

**Other information**:

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java [hunk 1]
```diff
@@ -390,79 +395,27 @@ private void calculateContingencySensitivityValues(PropagatedContingency conting
...
-                    newFactorStates = calculateFactorStates(loadFlowContext, factorGroups, newParticipatingElements);
-                }
                 // write contingency status
                 resultWriter.writeContingencyStatus(contingency.getIndex(), SensitivityAnalysisResult.Status.SUCCESS);
             } else {
                 // write contingency status
                 resultWriter.writeContingencyStatus(contingency.getIndex(), SensitivityAnalysisResult.Status.NO_IMPACT);
             }
 
+            // we need to recompute the factor states because the rhs has changed or
```
Comment (Hadrien-Godard) [inline src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:405]: Hello @p-arvy , please do not forget to complete this comment

Comment (p-arvy) [reply src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java]: Indeed. Ty for the comment @Hadrien-Godard.

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java [hunk 2]
```diff
@@ -340,18 +308,61 @@ private static double calculatePower(DcLoadFlowContext loadFlowContext, LfBranch
...
+    private void calculateSensitivityValuesForAContingency(DcLoadFlowContext loadFlowContext, OpenLoadFlowParameters lfParametersExt, SensitivityFactorHolder<DcVariableType, DcEquationType> validFactorHolder,
+                                                           SensitivityFactorGroupList<DcVariableType, DcEquationType> factorGroups, DenseMatrix factorStates, DenseMatrix contingenciesStates, DenseMatrix flowStates,
+                                                           PropagatedContingency contingency, Map<String, ComputedContingencyElement> contingencyElementByBranch,
+                                                           Set<LfBus> disabledBuses, List<ParticipatingElement> participatingElements, Set<String> elementsToReconnect,
+                                                           SensitivityResultWriter resultWriter, ReportNode reportNode, Set<LfBranch> partialDisabledBranches, boolean rhsChangedAfterConnectivityBreak) {
+        List<LfSensitivityFactor<DcVariableType, DcEquationType>> factors = validFactorHolder.getFactorsForContingency(contingency.getContingency().getId());
+        if (factors.isEmpty()) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:321]: @p-arvy and @Hadrien-Godard I did that, more coherent to me not to compute anything more if no factors for this contingency. Can you double-check?

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java [hunk 3]
```diff
@@ -340,18 +308,61 @@ private static double calculatePower(DcLoadFlowContext loadFlowContext, LfBranch
...
+            // get the lost phase tap changers for this contingency
+            Set<LfBranch> lostPhaseControllers = contingency.getBranchIdsToOpen().keySet().stream()
+                    .filter(element -> !elementsToReconnect.contains(element))
+                    .map(contingencyElementByBranch::get)
+                    .map(ComputedContingencyElement::getLfBranch)
+                    .filter(LfBranch::hasPhaseControllerCapability)
+                    .collect(Collectors.toSet());
+
+            // if a phase tap changer is lost or if the connectivity have changed, we must recompute load flows
+            if (!disabledBuses.isEmpty() || !lostPhaseControllers.isEmpty()) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:354]: Note that I have removed what you did new DisabledNetwork(disabledBuses, disabledPhaseTapChangers), I don't understand why we need it.

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java (hunk 2 above)
Comment (Hadrien-Godard) [reply src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:321]: Ok for me.

File: src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java (hunk 3 above)
Comment (Hadrien-Godard) [reply src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java:354]: Indeed, it is already covered in
DisabledNetwork disabledNetwork = new DisabledNetwork(disabledBuses, disabledBranches);

Thanks!

**Review by Hadrien-Godard — APPROVED** [review]

**Review by annetill — APPROVED** [review]
CI powsybl-open-rao passed, internal integration tests are not passing (one only and not a dc computation...). So I feel confident by merging this PR.

**Pull request merged**